conn.go

```diff
@@ -3,32 +3,46 @@
 package preconn
 
 import (
+	"bytes"
+	"errors"
+	"io"
 	"net"
 )
 
-// Conn is a net.Conn that supports replaying.
+// Conn is a net.Conn that supports replaying. Reads are not concurrency-safe.
 type Conn struct {
 	net.Conn
-	head []byte
+	head         io.Reader
+	consumedHead bool
 }
 
 // Wrap wraps the supplied conn and inserting the given bytes at the head of the
 // stream.
 func Wrap(conn net.Conn, head []byte) *Conn {
-	return &Conn{
-		Conn: conn,
-		head: head,
-	}
+	return WrapReader(conn, bytes.NewReader(head))
+}
+
+// WrapReader wraps the supplied conn, reading from 'head' first.
+func WrapReader(conn net.Conn, head io.Reader) *Conn {
+	return &Conn{conn, head, false}
 }
 
 // Read implements the method from net.Conn and first consumes the head before
 // using the underlying connection.
-func (conn *Conn) Read(b []byte) (int, error) {
-	n := copy(b, conn.head)
-	conn.head = conn.head[n:]
-	if n > 0 {
-		return n, nil
+func (conn *Conn) Read(b []byte) (n int, err error) {
+	if conn.consumedHead {
+		return conn.Conn.Read(b)
+	}
+	// "Read conventionally returns what is available instead of waiting for more."
+	// - https://golang.org/pkg/io/#Reader
+	// Thus if we read off conn.head, we return immediately rather than continuing to conn.Conn.
+	n, err = conn.head.Read(b)
+	if errors.Is(err, io.EOF) {
```

Comment:
The case of non EOF errors is a little interesting. I can't think of anything better to do than not handle it, but I thought I'd bring it up.

Reply:
Non-EOF errors are handled!  They'll be returned to the caller, same as if `conn.Conn.Read` returned an error.

Reply:
All we're doing here is `nil`ing EOF errors.  They don't truly represent an EOF as it pertains to the connection since there may be more data on `conn.Conn`.

Reply:
Yeah, I just meant that a non-EOF error on the head is an interesting/weird case that we don't have any special handling for in here (e.g. discard the rest of the head). I think keeping the head and just letting the caller deal with it as this code already does makes the most sense.

Reply:
Ah I see.  Agreed, I'm not sure what else we could do that would make sense.  Thanks!